File: r/src/hello_modules.r
library("R6")

HelloController <-
  R6Class("HelloController",
    private = list(
      greeting = "Hello"
    ),
    public = list(
      say = function(to) {
        print(paste(private$greeting, ", ", to, "!", sep = ""))
      }
    )
  )

parse_args = function() {
  args <- c(commandArgs(trailingOnly=TRUE)[1])
  names(args) <- c("world")
  return(args)
}

main = function() {
  args <- parse_args()
  controller <- HelloController$new()
  controller$say(args["world"])
}

main()

File: scala/src/HelloModules.scala
object HelloModules {
  def main(args: Array[String]): Unit = {
    var controller = new HelloController
    var parsedArgs = controller.parseArgs(args)
    controller.say(parsedArgs("world"))
  }
}

class HelloController {
  private var _greeting = "Hello"

  def parseArgs(args: Array[String]): Map[String, String] = {
    Map("world" -> args(0))
  }

  def say(to: String): Unit = {
    println(s"${_greeting}, ${to}!")
  }
}
